Add automate methods for VM import between providers

matobet opened the pull request:
Bug-Url: https://bugzilla.redhat.com/show_bug.cgi?id=1404920

Added automate methods required to back the transform vm service dialog (https://github.com/ManageIQ/manageiq/pull/13794 https://github.com/ManageIQ/manageiq-ui-classic/pull/383) facilitating v2v conversion of VMs between two infra providers.

The dialog consists of 3 select boxes (target infra provider, cluster, storage domain) and a single checkbox that controls whether thin provisioning is enabled. To supply those select boxes we provided the `list_infra_providers`, `list_clusters` and `list_storages` methods respectively.

matobet commented:
Depends on https://github.com/ManageIQ/manageiq/pull/13916

mkanoor commented:
@matobet 
Could you also add a spec for the automate methods in this PR

For the newer methods we are following a module approach so that we can test them outside of the Automate Engine. A good example to follow would be a method and spec like

Method

https://github.com/ManageIQ/manageiq-content/blob/master/content/automate/ManageIQ/Service/Generic/StateMachines/GenericLifecycle.class/__methods__/check_completed.rb

Spec

https://github.com/ManageIQ/manageiq-content/blob/master/spec/content/automate/ManageIQ/Service/Generic/StateMachines/GenericLifecycle.class/__methods__/check_completed_spec.rb

matobet commented:
@mkanoor Posted new (modularized) version of the scripts, can you please tell me if I'm on the right track?

mkanoor commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/import_infra_vm.rb:
```diff
@@ -0,0 +1,49 @@
+                  :storage_id => storage_id,
+                  :sparse     => sparse
+                )
+                exit MIQ_OK
```
@matobet 
We don't need the exit you can use @handle.root['ae_result'] = 'ok'

matobet replied:
done

mkanoor commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/import_infra_vm.rb:
```diff
@@ -0,0 +1,49 @@
+
+                provider.import_vm(
+                  vm.id,
+                  :name       => name,
```
@matobet 
Could we not use the local variables and directly use the attribute values from root
```
provider.import_vm(
@handle.root['vm'].id,
:name =>  @handle.root['dialog_name'],
:cluster_id => @handle.root['dialog_cluster'],
:storage_id => @handle.root['dialog_storage'],
:sparse => @handle.root['dialog_sparse'])
```

matobet replied:
done

matobet commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/list_infra_providers.rb:
```diff
@@ -0,0 +1,38 @@
+                values_hash = {}
+                values_hash[''] = '-- select target infrastructure provider from list --'
+
+                managers = @handle.vmdb('ManageIQ_Providers_Redhat_InfraManager').all.select(&:validate_import_vm)
```
@mkanoor When writing the specs for this script, how should I approach to mocking this call to `vmdb(ManageIQ_Providers_Redhat_InfraManager)`. I tried looking in other specs and found no example of this..

From what I understand the basic idea is to create the backing AR model, e.g. 

`let(:provider)    { FactoryGirl.create(:ems_redhat) }` 

and then create the "wrapping" 

`let(:svc_model_provider) { MiqAeMethodService::MiqAeServiceManageIQ_Providers_Redhat_InfraManager.find(provider.id) }` 

but I'm not sure how to proceed in this case... (tests are currently failing with `"Service Model not found"`.

matobet replied:
ok solved

mkanoor commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/list_clusters.rb:
```diff
@@ -0,0 +1,45 @@
+                  end
+                end
+                list_values = {
+                  'sort_by'   => :value,
```
@matobet 
Do you want the Dialogs to show items sorted by ID or by description/name?
If you want it to displayed sorted by name description the sort_by => 'description'

matobet replied:
done

mkanoor commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/list_clusters.rb:
```diff
@@ -0,0 +1,45 @@
+                    values_hash['!'] = '-- No clusters found --'
+                  else
+                    provider.ems_clusters.each do |cluster|
+                      values_hash[cluster.id.to_s] = cluster.name
```
@matobet 
could the key here be just id and not id string

matobet replied:
done

mkanoor commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/list_clusters.rb:
```diff
@@ -0,0 +1,45 @@
+                if !provider_id.nil? && !provider_id.empty? && provider_id != '!'
+                  provider = @handle.vmdb(:ext_management_system, provider_id)
+                  if provider.nil?
+                    values_hash['!'] = '-- No clusters found --'
```
@matobet 
We typically use nil => "None" as the default for 
https://github.com/ManageIQ/manageiq-content/blob/master/content/automate/ManageIQ/Cloud/Orchestration/Operations/Methods.class/__methods__/available_flavors.rb#L27

matobet replied:
done

mkanoor commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/list_clusters.rb:
```diff
@@ -0,0 +1,45 @@
+                values_hash = {}
+                values_hash[''] = '-- select cluster from list --'
+
+                provider_id = @handle.root['dialog_provider']
```
@matobet 
Are you expecting the @handle.root['dialog_provider'] to be ! at some point as a default value

matobet replied:
@mkanoor The `'!'` value of `provider_id` is being handled at line 17, if that is your question.

mkanoor commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/list_storages.rb:
```diff
@@ -0,0 +1,46 @@
+
+              def main
+                values_hash = {}
+                values_hash[''] = '-- select storage from list --'
```
@matobet 
I think the key here should be nil
``
value_hash[nil] = '-- select storage from list --'
``
On line 21 you are setting nil to None if no providers are found.

matobet replied:
done

mkanoor commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/list_clusters.rb:
```diff
@@ -0,0 +1,45 @@
+module ManageIQ
```
@matobet 
can you please add a spec for this method and all the other methods.
https://github.com/ManageIQ/manageiq-content/blob/master/spec/content/automate/ManageIQ/AutomationManagement/AnsibleTower/Operations/Methods.class/__methods__/available_credentials_spec.rb

matobet commented:
@mkanoor done, but now this patch also depends on the core patch adding required factories for the specs https://github.com/ManageIQ/manageiq/pull/14017

mkanoor commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/import_infra_vm.rb:
```diff
@@ -0,0 +1,48 @@
+                }
+
+                auto_approve = true
+                @handle.execute('create_automation_request', options, 'admin', auto_approve)
```
@matobet 
instead of admin you could use @handle.root['user'].userid

mkanoor replied:
@matobet 
Should this method be called CreateImportVMRequest

matobet replied:
@mkanoor ah, ok. I thought this has to be `'admin'` when I want to use the auto_approve feature - I understood this is the "approving" user being passed...

mkanoor replied:
@matobet This file should be deleted, since we swapped it out.

tinaafitz commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/create_vm_import_request.rb:
```diff
@@ -0,0 +1,46 @@
+#
```
@matobet Is there a spec test for this method?

matobet replied:
done

tinaafitz commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/list_clusters.rb:
```diff
@@ -0,0 +1,45 @@
+                values_hash[nil] = '-- select cluster from list --'
+
+                provider_id = @handle.root['dialog_provider']
+                if !provider_id.nil? && !provider_id.empty? && provider_id != '!'
```
@matobet you could use if provider.present? instead of !provider_id.nil? && !provider_id.empty?

matobet replied:
done

tinaafitz commented on content/automate/ManageIQ/Infrastructure/VM/Transform/StateMachines/VmImport.class/__methods__/submit_vm_import.rb:
```diff
@@ -0,0 +1,40 @@
+              def main
+                provider = @handle.vmdb(:ext_management_system, @handle.root['provider_id'])
+                @handle.log(:info, 'Submitting Import')
+                new_ems_ref = provider.submit_import_vm(
```
@matobet Where is submit_import_vm defined?

matobet replied:
@tinaafitz Defined in https://github.com/ManageIQ/manageiq-providers-ovirt/pull/21 and exposed in https://github.com/ManageIQ/manageiq-automation_engine/pull/4

matobet commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/import_vm.yaml:
```diff
@@ -0,0 +1,10 @@
+---
```
I'm sorry but when I try removing this file, I start getting this error:
```
[----] I, [2017-05-15T14:33:16.784699 #5502:1097120]  INFO -- : Q-task_id([resource_action_877]) Updated namespace [miqaedb:/System/Request/Call_Instance#create  ManageIQ/System]
[----] I, [2017-05-15T14:33:16.793390 #5502:1097120]  INFO -- : Q-task_id([resource_action_877]) Following Relationship [miqaedb:/ManageIQ/Infrastructure/VM/Transform/Import/ImportVm#create]
[----] I, [2017-05-15T14:33:16.812444 #5502:1097120]  INFO -- : Q-task_id([resource_action_877]) Instance [/ManageIQ/Infrastructure/VM/Transform/Import/ImportVm] not found in MiqAeDatastore - trying [.missing]
[----] E, [2017-05-15T14:33:16.812550 #5502:1097120] ERROR -- : Q-task_id([resource_action_877]) Instance [/ManageIQ/Infrastructure/VM/Transform/Import/ImportVm] not found in MiqAeDatastore
[----] I, [2017-05-15T14:33:16.812592 #5502:1097120]  INFO -- : Q-task_id([resource_action_877]) Followed  Relationship [miqaedb:/ManageIQ/Infrastructure/VM/Transform/Import/ImportVm#create]
[----] I, [2017-05-15T14:33:16.812857 #5502:1097120]  INFO -- : Q-task_id([resource_action_877]) Followed  Relationship [miqaedb:/System/Request/Call_Instance#create]
```

tinaafitz replied:
@matobet, these two yaml files create the /ManageIQ/Infrastructure/VM/Transform/Import/ImportVm instance:
content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/importvm.yaml
content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/import_vm.yaml

You should be able to remove:
content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/import_vm.yaml
and the instance will be created properly by:
content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/importvm.yaml

matobet replied:
@tinaafitz done, removed the importvm one

matobet commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/list_storages.rb:
```diff
@@ -0,0 +1,46 @@
+
+                provider_id = @handle.root['dialog_provider']
+                @handle.log(:info, "Selected provider: #{provider_id}")
+                if !provider_id.nil? && !provider_id.empty? && provider_id != '!'
```
done

matobet commented on spec/content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/import_infra_vm_spec.rb:
```diff
@@ -0,0 +1,38 @@
+require_domain_file
```
yes, thank you, modified to test `create_vm_import_request`

matobet commented on content/automate/ManageIQ/System/Event/EmsEvent/RHEVM.class/__methods__/update_vm_import_status.rb:
```diff
@@ -0,0 +1,13 @@
+#
```
@tinaafitz actually here I would most likely appreciate a bit of guidance. Since I appear to be the first adding a method to an Event class in the `System` namespace I could really use an example of how such a test should be structured....

tinaafitz replied:
@matobet, Here's a spec test I created for the event class method. It needs work, but it has the basics. Also, I had to modify the Automate method for the spec tests to run properly, it probably needs work/testing as well.  I hope this helps. 

reformatted method
`#
# Description: Update VM import status.
#

module ManageIQ
  module Automate
    module System
      module Event
        module EmsEvent
          module RHEVM
            class UpdateVmImportStatus
              def initialize(handle = $evm)
                @handle = handle
              end

              def main
                update_import_status(vm)
              end

              private

              def vm_id
                event_stream = @handle.root['event_stream']
                if event_stream.nil?
                  @handle.log(:error, 'event_stream not found')
                  raise 'event_stream not found'
                end
                event_stream.vm_or_template_id
              end

              def vm
                @handle.vmdb('Vm', vm_id).tap do |vm|
                  if vm.nil?
                    @handle.log(:error, 'VM object not found')
                    raise 'VM object not found'
                  end
                end
              end

              def import_status
                @handle.root['event_type'] == 'IMPORTEXPORT_IMPORT_VM_FAILED' ? 'failure' : 'success'
              end

              def update_import_status(vm)
                vm.custom_set(:import_status, import_status)
                exit MIQ_OK
              end
            end
          end
        end
      end
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  ManageIQ::Automate::System::Event::EmsEvent::
    RHEVM::UpdateVmImportStatus.new.main
end
`
`require_domain_file

describe ManageIQ::Automate::System::Event::EmsEvent::
          RHEVM::UpdateVmImportStatus do

  let(:user)          { FactoryGirl.create(:user_with_group) }
  let(:miq_server)    { EvmSpecHelper.local_miq_server }
  let(:ems)           { FactoryGirl.create(:ems_vmware_with_authentication, :zone => FactoryGirl.create(:zone)) }
  let(:vm)            { FactoryGirl.create(:vm_vmware, :ext_management_system => ems) }
  let(:miq_event)     { FactoryGirl.create(:miq_event, :event_type => "some_event") }

  let(:ems_event) do
    FactoryGirl.create(:ems_event, :vm_or_template => vm, :ext_management_system => ems)
  end
  let(:svc_model_miq_server) { MiqAeMethodService::MiqAeServiceMiqServer.find(miq_server.id) }
  let(:svc_model_user) { MiqAeMethodService::MiqAeServiceUser.find(user.id) }

  let(:root_hash) do
    { 'event_stream' => MiqAeMethodService::MiqAeServiceEmsEvent.find(ems_event.id),
      'event_type'   => 'IMPORTEXPORT_IMPORT_VM_FAILED' }
  end

  let(:ae_service) do
    Spec::Support::MiqAeMockService.new(root_object).tap do |service|
      current_object = Spec::Support::MiqAeMockObject.new
      current_object.parent = root_object
      service.object = current_object
    end
  end

  let(:root_object) do
    obj = Spec::Support::MiqAeMockObject.new(root_hash)
    obj["user"] = svc_model_user
    obj["miq_server"] = svc_model_miq_server
    obj
  end

  it "method succeeds" do
    described_class.new(ae_service).main
  end
end
`

tinaafitz replied:
@matobet, Here's a spec test I created for the event class method. It needs work, but it has the basics. Also, I had to modify the Automate method for the spec tests to run properly, it probably needs work/testing as well. I hope this helps.
`require_domain_file

describe ManageIQ::Automate::System::Event::EmsEvent::
          RHEVM::UpdateVmImportStatus do

  let(:user)          { FactoryGirl.create(:user_with_group) }
  let(:miq_server)    { EvmSpecHelper.local_miq_server }
  let(:ems)           { FactoryGirl.create(:ems_vmware_with_authentication, :zone => FactoryGirl.create(:zone)) }
  let(:vm)            { FactoryGirl.create(:vm_vmware, :ext_management_system => ems) }
  let(:miq_event)     { FactoryGirl.create(:miq_event, :event_type => "some_event") }

  let(:ems_event) do
    FactoryGirl.create(:ems_event, :vm_or_template => vm, :ext_management_system => ems)
  end
  let(:svc_model_miq_server) { MiqAeMethodService::MiqAeServiceMiqServer.find(miq_server.id) }
  let(:svc_model_user) { MiqAeMethodService::MiqAeServiceUser.find(user.id) }

  let(:root_hash) do
    { 'event_stream' => MiqAeMethodService::MiqAeServiceEmsEvent.find(ems_event.id),
      'event_type'   => 'IMPORTEXPORT_IMPORT_VM_FAILED' }
  end

  let(:ae_service) do
    Spec::Support::MiqAeMockService.new(root_object).tap do |service|
      current_object = Spec::Support::MiqAeMockObject.new
      current_object.parent = root_object
      service.object = current_object
    end
  end

  let(:root_object) do
    obj = Spec::Support::MiqAeMockObject.new(root_hash)
    obj["user"] = svc_model_user
    obj["miq_server"] = svc_model_miq_server
    obj
  end

  it "method succeeds" do
    described_class.new(ae_service).main
  end
end



`

tinaafitz replied:
@matobet 
`#
# Description: Update VM import status.
#

module ManageIQ
  module Automate
    module System
      module Event
        module EmsEvent
          module RHEVM
            class UpdateVmImportStatus
              def initialize(handle = $evm)
                @handle = handle
              end

              def main
                update_import_status(vm)
              end

              private

              def vm_id
                event_stream = @handle.root['event_stream']
                if event_stream.nil?
                  @handle.log(:error, 'event_stream not found')
                  raise 'event_stream not found'
                end
                event_stream.vm_or_template_id
              end

              def vm
                @handle.vmdb('Vm', vm_id).tap do |vm|
                  if vm.nil?
                    @handle.log(:error, 'VM object not found')
                    raise 'VM object not found'
                  end
                end
              end

              def import_status
                @handle.root['event_type'] == 'IMPORTEXPORT_IMPORT_VM_FAILED' ? 'failure' : 'success'
              end

              def update_import_status(vm)
                vm.custom_set(:import_status, import_status)
                exit MIQ_OK
              end
            end
          end
        end
      end
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  ManageIQ::Automate::System::Event::EmsEvent::
    RHEVM::UpdateVmImportStatus.new.main
end
`

matobet replied:
@tinaafitz thank you very much for your help! With this example I've been able to write the tests for this method, posted new version.

matobet commented on content/automate/ManageIQ/Infrastructure/VM/Transform/StateMachines/VmImport.class/__methods__/submit_vm_import.rb:
```diff
@@ -0,0 +1,40 @@
+              end
+
+              def main
+                provider = @handle.vmdb(:ext_management_system, @handle.root['provider_id'])
```
Well this code is mostly forwarding all arguments to the import_vm backend where all necessary validation should be performed. I am not entirely sure if duplicating it here is worth it...

matobet commented:
@tinaafitz pushed new version, added missing tests and hopefully addressed your comments :)

matobet commented on content/automate/ManageIQ/Infrastructure/VM/Transform/StateMachines/VmImport.class/__methods__/wait_for_vm_import.rb:
```diff
@@ -0,0 +1,52 @@
+#
```
@tinaafitz done

tinaafitz commented on content/automate/ManageIQ/Infrastructure/VM/Transform/StateMachines/VmImport.class/__class__.yaml:
```diff
@@ -0,0 +1,53 @@
+---
```
@matobet We're working on documenting guidelines for adding State Machines to the ManageIQ domain. 

Can you modify your ImportVM State Machine as follows:

Minimum Requirements for ManageIQ domain State Machines:
(In addition to the specific processing states required by the new state machine)   

1. Minimum of 3 Pre states(pre1, pre2, pre3)
2. Minimum of 3 Post states(post1, post2, post3)

3. update_*_status method required for every states on_entry, on_exit, on_error
 a. * update request with enhanced message including 
       updated_message  = "[#{$evm.root['miq_server'].name}] "
       updated_message += "Step [#{$evm.root['ae_state']}] "
       updated_message += "Status [#{status}] "
       updated_message += "Message [#{prov.message}] "
       updated_message += "Current Retry Number [#{$evm.root['ae_state_retries']}]" if 
            $evm.root['ae_result'] == 'retry'
 b. * set user_message in request:
       prov.miq_request.user_message = updated_message

 c. * create notification and error log message on error
      if $evm.root['ae_result'] == "error"
          $evm.create_notification(:level => "error" \
                            :message => "VM Provision Error: #{updated_message}")
          $evm.log(:error, "VM Provision Error: #{updated_message}")
      end

4. Finished state 
    Can use CommonMethods finished method: 
    /System/CommonMethods/StateMachineMethods/your_instance_name_here
    (Add an instance to the /System/CommonMethods/StateMachineMethods class with your      instance name. Can optionally modify task_finished method in that class for special processing)

matobet replied:
@tinaafitz ok, 1) and 2) seems pretty straightforward (I assume the states to be of type `string`) 
but regarding 3) what is exactly necessary? add 3 *methods* for each state of the state machine (with the new pre and post that would total to 8*3=24 new methods)?

matobet replied:
@tinaafitz and regarding 4) the state will be part of the ImportVm state machine (after the last "post" state)?

matobet replied:
@tinaafitz Done, please review at your nearest convenience :)

tinaafitz commented on content/automate/ManageIQ/Infrastructure/VM/Transform/Import.class/__methods__/create_vm_import_request.rb:
```diff
@@ -0,0 +1,46 @@
+              end
+
+              def main
+                options = {
```
@matobet - Could you add basic validation that the required arguments exist before calling execute to create the automation request?

Here's a sample basic validation method:
def validate_root_args(arg_names)
  arg_names.each do |name|
    unless $evm.root[name].present?
      msg = "Error, required root attribute: #{name} not found"
      $evm.log(:error, msg)
      raise msg
    end
  end
end

and you can call it prior to calling execute....

validate_root_args(%w(vm dialog_name dialog_provider dialog_cluster dialog_sparse))

matobet replied:
done

tinaafitz commented on content/automate/ManageIQ/Infrastructure/VM/Transform/StateMachines/VmImport.class/__methods__/submit_vm_import.rb:
```diff
@@ -0,0 +1,40 @@
+
+              def main
+                provider = @handle.vmdb(:ext_management_system, @handle.root['provider_id'])
+                @handle.log(:info, 'Submitting Import')
```
@matobet - Could you add basic validation that the required arguments exist before calling submit_import_vm?

Here's a sample basic validation method:
def validate_root_args(arg_names)
  arg_names.each do |name|
    unless $evm.root[name].present?
      msg = "Error, required root attribute: #{name} not found"
      $evm.log(:error, msg)
      raise msg
    end
  end
end

and you can call it prior to calling execute....

validate_root_args(%w(vm name cluster_id storage_id sparse))

matobet replied:
done

matobet commented on spec/content/automate/ManageIQ/System/Event/EmsEvent/RHEVM.class/__methods__/update_vm_import_status_spec.rb:
```diff
@@ -0,0 +1,54 @@
+  let(:miq_server) { EvmSpecHelper.local_miq_server }
+  let(:ems)        { FactoryGirl.create(:ems_redhat) }
+  let(:vm)         { FactoryGirl.create(:vm_redhat) }
+  let(:miq_event)  { FactoryGirl.create(:miq_event, :event_type => event_type) }
```
done

matobet commented:
Hi @tinaafitz thank you for the review, I addressed all your comments but I'm not sure what exactly do you mean by *setup/prerequisites/use cases for documentation purposes*?

tinaafitz commented:
@matobet Thanks for making the changes. I'll ask @mkanoor to review.
You probably already have all of the necessary documentation information in your PRs.
The purpose is to answer the question of how do I use this feature? Although I've reviewed your PR, I wouldn't know how best to guide someone who wanted to use this new feature. 

I think the information is all here, maybe if you could provide a brief explanation to a new user wanting to use v2.
https://docs.google.com/document/d/1TnAjYJd041vCWu71ks4oxT_jjSrlDE2nZWLBXK4ZKkg/edit
https://trello.com/c/uh3V62OZ/11-v2v-in-miq

tinaafitz commented:
@mkanoor @matobet has made all of the requested changes.  Please review.

mkanoor commented:
@matobet Can you fix the rubocop warning, we can use a next if the variable is present else raise an error

mkanoor commented:
@matobet 
```
              def validate_root_args(arg_names)
                arg_names.each do |name|
                  next if @handle.root[name].present?
                  raise "Error, required root attribute: #{name} not found"
                 end
              end
```

matobet commented:
@mkanoor done

matobet commented:
@tinaafitz as per our discussion, posting here short feature description:

# Feature Overview
- Allow migration of user workloads (VMs) from VMware infra provider to RHV provider
- So far bulk operations not supported, need to  individual VMs one-by-one
- User chooses target (RHV) provider, cluster and storage domain
- So far only the Vmware -> RHV is direction is supported (no other permutations of source & destination)

# REST
Added a new action to the providers REST resource that allows for importing vms.
`POST /api/providers/:id`
```json
{
  "action" : "import_vm", 
  "resource" : {
      "source": { "href": "http://localhost:3000/api/vms/:source_vm_id" },
      "target": {
          "name": "<name of newly created VM in RHEV>",
          "data_store": { "href": "http://localhost:3000/api/:target_data_store_id" },
          "cluster": { "href": "http://localhost:3000/api/:target_cluster_id" },
          "sparse": "[true|false]"
      }
  }
}
```
For now only the combination of a Vmware source VM and a RHEV as the target provider is supported.

# UI
- New button in the “vm detail” (“Transform this VM to RHV”) 
![transform-vm-to-rhv-button](https://cloud.githubusercontent.com/assets/1173428/26361507/83aa24ae-3fdb-11e7-8db0-8644aa370dc6.png)


- It should be visible only for VMware VMs
- It should be enabled (not greyed out) only when there exists a suitable target provider in the system (in this case RHV infra provider of version 4.0+)
- Spawns dialog with the following options 
![transform-vm-dialog_](https://cloud.githubusercontent.com/assets/1173428/26361515/89e93c24-3fdb-11e7-9a31-99a00b0285a8.png)
  - Name
  - Cluster
  - Storage domain
  - Thin provisioning (yes/no checkbox)
For the created VM.
- After submission the status and progress of the operation should be tracked along with other automate requests.

mkanoor commented:
@matobet Looks good.
👍

simaishi commented:
Fine backport details:

```
$ git log -1
commit 868138f7b46eb73d639d8c8bed337cea78da34c7
Author: Greg McCullough <gmccullo@redhat.com>
Date:   Wed May 24 08:28:35 2017 -0400

    Merge pull request #36 from matobet/v2v-automate
    
    Add automate methods for VM import between providers
    (cherry picked from commit bd5bb97383ea203cc2bde9d740a09624a08e01f2)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1459996
```
